analyzers/src/SonarAnalyzer.CSharp/Rules/DeclareParameterBeforeUsage.cs

```diff
@@ -0,0 +1,131 @@
+                if (cs.Compilation.GetTypeByMetadataName(KnownType.Microsoft_AspNetCore_Components_ParameterAttribute) is null)
+                {
+                    return;
+                }
+
+                cs.RegisterNodeAction(c =>
+                {
+                    var method = (MethodDeclarationSyntax)c.Node;
+                    if (!IsBuildRenderTreeMethod(method))
+                    {
+                        return;
+                    }
+
+                    ITypeSymbol currentComponent = null;
+                    Dictionary<string, ComponentDescriptor> descriptors = new();
+                    foreach (var invocation in method.DescendantNodes().OfType<InvocationExpressionSyntax>())
+                    {
+                        var targetMethod = c.SemanticModel.GetOperation(invocation).AsInvocation().Value.TargetMethod;
+                        if (targetMethod.Name.Equals("OpenComponent")
+                            && targetMethod.TypeArguments is { Length: 1 } arguments
+                            && arguments[0].BaseType.Is(KnownType.Microsoft_AspNetCore_Components_ComponentBase))
+                        {
+                            currentComponent = targetMethod.TypeArguments[0];
+                        }
+                        else if (currentComponent != null
+                            && targetMethod.Name.Equals("AddAttribute")
+                            && targetMethod.Parameters is { Length: >= MinAddAttributeParameters } parameters
+                            && parameters[1].Type.Is(KnownType.System_String)
+                            && invocation.ArgumentList.Arguments[1].Expression.StringValue(c.SemanticModel) is { } parameterName
+                            && GetComponentDescriptor(currentComponent, descriptors) is var descriptor
+                            && !descriptor.HasMatchUnmatchedParameters
+                            && !descriptor.Parameters.Contains(parameterName))
+                        {
+                            var location = c.Tree.FilePath.EndsWith("razor.g.cs", StringComparison.OrdinalIgnoreCase)
+                                ? null // The diagnostic is reported at the beginning of the file because the attribute location cannot be mapped back.
+                                : invocation.GetLocation();
+                            c.ReportIssue(Diagnostic.Create(Rule, location, parameterName, currentComponent.Name));
+                        }
+                        else if (targetMethod.Name.Equals("CloseComponent"))
+                        {
+                            currentComponent = null;
+                        }
+                    }
+                }, SyntaxKind.MethodDeclaration);
+            });
+
+    private static bool IsBuildRenderTreeMethod(MethodDeclarationSyntax method) =>
+        method.NameIs("BuildRenderTree")
+        && method.ParameterList.Parameters.Count == 1
+        && method.Modifiers.Any(SyntaxKind.OverrideKeyword);
```

Comment:
You are right that we are not raising for invocation outside of the `BuildRenderTree` method, but the snipped you provided is raising correctly since we look for all descendant `InvocationExpressionSyntax` (nested or non nested). Do you have some generated snippets that are invoking AddAttribute outside the BuildRenderTree?

Reply:
It won't raise in cases like [this one](https://github.com/MudBlazor/MudBlazor/blob/59f7e62b40fbc015ad0e4eb0cf718d7031c9b8e7/src/MudBlazor/Services/DialogService.cs#L38-L44) or [that one](https://github.com/MudBlazor/MudBlazor/blob/59f7e62b40fbc015ad0e4eb0cf718d7031c9b8e7/src/MudBlazor/Services/DialogService.cs#L95-L128).

Those cases are uncommon; I will let you decide whether those FNs are worth fixing.